Objecttype

tardyp opened the pull request:
This is first baby step to implement the raml type system:

Type system base framework, which for now only supports the object type.

documentation is describing how I think it should look like.

I documented the validate() function, which I did not implement yet.
Let me know if you like how it works.

matiasb committed:
Fixed how args are passed to pytest in setup.py.

econchick committed:
Fix for issue 47

econchick committed:
Update changelog for v0.1.9

econchick committed:
Reorg of parsing & util logic to remove duplication and prepare for
RAML v1.0 features.

Commit:
extract raml version from the header

pyYaml does not allow extraction of the comments so this is a little bit hacky

Commit:
add validation of the version allowed

econchick committed:
Cleaned up Python version compatibility and minor test breakage

Commit:
forgot to commit a test raml

Commit:
types: base framework for type system

Implement the 'object' base type

bump the required version of attrs in order to get the 'convert' feature

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+from six import iteritems
+
+from ..errors import UnknownTypeError
+from ..parameters import Content
```
The proper way to import (for compatibility) is `from ramlfications.parameters import Content`

tardyp replied:
ok

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+class Property(object):
+    required                 = attr.ib(default=None, repr=False)
+    default                  = attr.ib(default=None, repr=False)
+    type                     = attr.ib(default="string")
```
I'd say too many spaces between these; I'd go with the same way you addressed the spacing under `class ObjectType(BaseType)` below

econchick commented on requirements.txt:
```diff
@@ -3,7 +3,7 @@ markdown2==2.3.0
 termcolor==1.1.0
 six==1.8.0
-attrs==15.0.0
+attrs==15.2.0
```
What does the updated version give us?

tardyp replied:
it gives us the convert parameter for attr.ib

econchick commented on tests/data/validate/invalid-version.raml:
```diff
@@ -0,0 +1,2 @@
+#%RAML 0.9
+title: Spotify Web API
```
I merged your last PR, so I think a rebase is necessary.

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+class BaseType(object):
+    name            = attr.ib()
+    description     = attr.ib(default="", repr=False, convert=Content)
+    type            = None
```
Another comment: `type` (throughout this whole module) is confusing and contradictory to the `type` property in `ResourceNode` and `ResourceTypeNode`. Maybe `data_type`?

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2015 Spotify AB
```
I'm _pretty_ sure you're not a Spotify employee :) I think, since this is a new module, this should be `2016` and your name instead of Spotify.

tardyp replied:
I'm not a big fan of putting my name in source code. This file will obviously be modified by lot of other contributors after me, as ramlfication is meant for great success.
In buildbot, we usually put "Copyright The Buildbot Developers"

Would you agree to put "Copyright The Ramlfications developers"

econchick commented on tests/raml10tests/test_types.py:
```diff
@@ -0,0 +1,34 @@
+
+# as much as possible, those tests are implementing the spec
+# the filenames are mapped against
+def loadapi(fn):
```
perhaps make use of `pytest.fixture`? Used in other tests in this repo too.

tardyp replied:
ok. I'm not yet very familiar to pytest, so I need to see how that could work.

econchick commented on ramlfications/raml.py:
```diff
@@ -30,6 +30,7 @@ class RootNode(object):
+    :param str raml_version: RAML spec version
     :param str base_uri: API's base URI
     :param list base_uri_params: parameters for base URI, or ``None``. \
         The order of ``base_uri_params`` will follow the order \
```
Don't forget to add the param description for the new `types` attribute

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+
+    # spec:
+    # string is default type when nothing else defined
+    typeexpr = raml_def.pop('type', 'string')
```
is this `copy` and `pop` needed? can't you just do a dictionary lookup?  I'd like to avoid `copy` because it tends to add a lot of overhead, especially with super-long RAML files like `ramlfications/tests/data/examples/twitter.raml` and `ramlfications/tests/data/examples/github.raml`.

tardyp replied:
its not a deep copy, just a copy of the type the goal of this dict is just to be passed as keyword to the constructor of the type class.
as 'type' is not an attr.ib of the class then we need to pop it.
so that is a small copy of 5-10 references

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+def type_class(klass):
+    if klass.type is not None:
+        __type_registry[klass.type] = klass
+    return klass
```
Interesting way to handle this!

tardyp replied:
its the registry design pattern which is used in twisted/zope interfaces in a couple of places.

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+
+def type_class(klass):
+    if klass.type is not None:
+        __type_registry[klass.type] = klass
```
What happens if there are two of the same class types? Here, it'll just overwrite. But if I'm thinking about future-proofing, would there ever be the case where someone uses this decorator without realizing the name of a class is already defined?

econchick replied:
Hmm it looks like `Property` and `StringType` would overwrite each other.

tardyp replied:
I think we should raise in that case indeed. I should change the name of attribute from 'type' to 'typename' to be more explicit.
The goal of this registry is to combine types defined by the spec (object, string) and types defined by the user's raml.

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+        raise UnknownTypeError("{0} type expression is not supported or not defined".format(typeexpr))
+
+@attr.s
+class BaseType(object):
```
Can I have doc strings for all these public classes & functions?

econchick replied:
For the base class, it seems to be missing `schema`, `example`, `examples`, `display_name`, and `annotation_name` (which could just be a place holder right now since annotation types haven't been implemented yet)

tardyp replied:
indeed, I forgot to update the docstrings

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+@attr.s
+class BaseType(object):
+    name            = attr.ib()
+    description     = attr.ib(default="", repr=False, convert=Content)
```
Since the RAML spec says that this will be a markdown string, I would encourage using the same pattern of the rest of the package. I have a [`desc` defined](https://github.com/spotify/ramlfications/blob/master/ramlfications/parameters.py#L86) and then a [`description` property](https://github.com/spotify/ramlfications/blob/master/ramlfications/parameters.py#L104-L108) to handle the parsing of markdown.  It may not be the best approach (open to ideas) but since it's expected to be markdown, we should handle that.

econchick replied:
Oh ha! I just noticed that `convert`! Is that what's added in the updated attrs?! That's pretty awesome...

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+class ObjectType(BaseType):
+    type                 = "object"
+    properties           = attr.ib(default=None, convert=parse_properties)
+    minProperties        = attr.ib(repr=False, default=0)
```
The casing is inconsistent compared to the rest of the module. Every `attr` is snake case (and abbreviated but not a hard rule), e.g. `min_prop` or `min_properties`

tardyp replied:
it has to be that way if we want to use the attrs generated constructor to directly take the yaml dict.

as the raml says camlCase I would expect we use camlCase for our object representations of the raml

econchick commented on ramlfications/parser/types.py:
```diff
@@ -0,0 +1,67 @@
+@attr.s
+class ObjectType(BaseType):
+    type                 = "object"
+    properties           = attr.ib(default=None, convert=parse_properties)
```
omg `convert` is freaking awesome. This makes me want to see where else I can use it!

econchick commented on ramlfications/errors.py:
```diff
@@ -63,3 +63,7 @@ class MediaTypeError(Exception):
     pass
+
+
+class UnknownTypeError(Exception):
```
This is a bit confusing, could we use `UnknownDataTypeError`?  Also I think this should inherit from `BaseRAMLError` like the other RAML-specific errors.

tardyp replied:
oh yes, I think this is a bad copy paste

econchick commented on tests/raml10tests/test_types.py:
```diff
@@ -0,0 +1,34 @@
+
+
+# Security scheme properties:
+# name, raw, type, described_by, desc, settings, config, errors
```
copy pasta?

econchick commented on ramlfications/parser/main.py:
```diff
@@ -73,6 +74,13 @@ def schemas():
+        _types = _get(raml, "types")
+        if not _types:
+            return None
+        return dict([(k, create_type(k, v))
```
I would like to preserve order here - to mimic what is in a RAML file - so perhaps create/return an OrderedDict.

tardyp replied:
ok

econchick commented on docs/extendedusage.rst:
```diff
@@ -198,6 +200,23 @@ Traits
+.. code-block:: python
+
+    >>> api.types
+    {'Person': ObjectType(name='Person', properties={'name': Property(type='string')})}
```
I like the `ObjectType` representation.

econchick replied:
Hmm but I'm not sure a dictionary is appropriate. The rest of the library returns a list (which preserves order) of the relevant objects, like `api.resources` and `resource.query_params` etc.

tardyp replied:
what I don't like with OrderedDict is the poor concision of its **repr**
I think a key feature of ramlfication is its nice **repr** of the parsed raml, which allows people to write their own parsers with print and REPL debugging.
My followup patch will contain a custom OrderedDict with a **repr** which is very similar as dict, only prefixed with 'o'

 "o{'Person': ObjectType(name='Person', properties=o{'name': Property(type='string')})}"

econchick commented on docs/extendedusage.rst:
```diff
@@ -198,6 +200,23 @@ Traits
+    >>> api.types
+    {'Person': ObjectType(name='Person', properties={'name': Property(type='string')})}
+    >>> person = api.types['Person']
+    >>> person.type
```
hmmmm this would actually make me thing that `person` is inheriting from another data type here.

tardyp replied:
well this is actually the case, its inheriting from the base type "object"
I need to think a little bit more on how the type inheritance should be described

econchick commented on docs/extendedusage.rst:
```diff
@@ -198,6 +200,23 @@ Traits
+    'a Person is a type describing human beings'
+    >>> person.properties
+    {'name': Property(type='string')})
+    >>> person.validate({'foo': 'bar'})
```
wouldn't it actually be `validate(raml_file)`? Here this implies that there's a `validate` function attached to the object, which isn't true (and I wouldn't want it).

tardyp replied:
What I mean here is to use the raml spec as a data validation mean
validate() takes an object and raise ValidationError if the object is not matching the type specified by the raml.

The background here is that my team is using http://python-eve.org/ in order to implement simple CRUD microservices backends.
python-eve is using Cerberus: http://python-eve.org/validation.html as its default data validation framework, but for me it would make a lot of sense to use raml and ramlfication instead. python-eve is already very modular in that matter.

Obviously this part is more for medium term

econchick commented on docs/extendedusage.rst:
```diff
@@ -198,6 +200,23 @@ Traits
+    >>> person.description
+    'a Person is a type describing human beings'
+    >>> person.properties
+    {'name': Property(type='string')})
```
Same argument of `list` vs `dict` as the above `api.types`

tardyp replied:
as everywhere else, types are referred by name, I think it is better to store it as a dict, which has fast and easy 'byName' access. I would have the same comment for ResourceType implementation, which is afaik a list

econchick commented:
This is awesome! A few more thoughts:
- It's a bit inconsistent with the rest of the package to define the type object "model" in the same module as the parser. I'd like to see it in the `raml.py` module. I'd also be open to breaking up the `raml.py` module similar to the `utils` and `parser` directory with a `raml` directory. 
- I need to do this for the rest of the package, but I've started thinking about how to handle when attributes are optional, denoted by a `?`. Right now, I only have optionality implemented for resource type methods.  I think I may end up making an API-breaking change since `resource_type.optional` wouldn't make sense any longer. I'm also thinking a `convert` from attrs or a decorator may be the way to go.  So, just something to keep in the back of your mind.

I really like your style and the approach; some clever things (the good kind of clever!). Hopefully I'm not too nit picky or critical. And definitely let me know if the new organization of the package doesn't make sense.

econchick commented:
Also - I just created a [new branch](https://github.com/spotify/ramlfications/tree/v0.2.0-dev) on the Spotify repo. I cleaned up a lot of the commit history.  This is where you can create your PRs against now :)

econchick commented:
Also - what do you think of the expected behavior I detailed a month or so ago [here](https://github.com/spotify/ramlfications/wiki/RAML-v1.0:-Data-Types#expected-output)?

tardyp commented:
Here is the spotify version of the PR.
https://github.com/spotify/ramlfications/pull/80

yes, I really appreciate the detailled review and critics, especially as the TZ does not help much to discuss design I think best approach is that I upload code often so that we are inline.

The codestyle of ramlfication is sometimes different from the one use use in buildbot (for example you use a lot of functions for the core logic while bbot is very object oriented)
I apologize if sometimes I make mistake. Obviously I need to adapt to the upstream codestyle, even if I may try to push some of my own design pattern if I think it is appropriate.

About the expected output, I actually forgot about it, and started with my own ideas.
for the data_type being a object rather than an identifier, I was actually starting with the second option, but maybe the pros of duplicating the references actually make sense especially as data_type is a type expression, and there can also be inline type definition.

econchick committed:
add travis config

econchick committed:
fixed order of uri/base uri params; create undeclared uri params

econchick committed:
Merge pull request #82 from spotify/lynn/fix-uri-params

fixed order of uri/base uri params; create undeclared uri params

econchick committed:
uncommented out forgotten but working tests

econchick committed:
remove random notes file that was not meant to be commited

econchick committed:
remove unneeded commented-out code

Commit:
Raml parser: support parsing fragments as the root node

For now, only datatype is introduced (actual implementation will be in the data type PR)

raml1.0 introduces possibility to split your REST api def into several parts

Interresting side effect of this is that you can use raml as a data-validator library.

a raml datatype file can be used for the same means as:
json schema
http://cerberus.readthedocs.org
https://pypi.python.org/pypi/yamltypes

Commit:
add support for int types

Commit:
add support for bool

Commit:
p3k support

Commit:
boolean is scalar, not number

Signed-off-by: Pierre Tardy <pierre.tardy@intel.com>

Commit:
more camlcase fixes

econchick committed:
Merge pull request #83 from tardyp/ramlfragments

Raml parser: support parsing fragments as the root node
